lib/service/service.go

```diff
@@ -5254,17 +5267,49 @@ func (process *TeleportProcess) initSecureGRPCServer(
 			authMiddleware.StreamInterceptor(),
 		),
 		grpc.Creds(credentials.NewTLS(
-			copyAndConfigureTLS(serverTLSConfig, process.log, authMiddleware.AccessPoint, conn.ServerIdentity.ClusterName),
+			copyAndConfigureTLS(serverTLSConfig, process.log, cfg.accessPoint, clusterName),
 		)),
 	)
 
+	process.OnExit("proxy.grpc.secure", func(interface{}) {
+		server.GracefulStop()
+	})
+
 	process.RegisterCriticalFunc("proxy.grpc.secure", func() error {
-		process.log.Infof("Starting proxy gRPC server on %v.", listener.Addr())
-		return trace.Wrap(server.Serve(listener))
+		kubeServer, err := kubegprc.New(kubegprc.Config{
```

Comment:
Is there a reason that in `initPublicGRPCServer` we register the gRPC service onto the server immediately after instantiating it, but here, we are placing this into `RegisterCriticalFunc` and essentially registering the gRPC service onto the server just before we start listening ? It might seem like a silly question but I think we ought to try and be consistent where we can.